itest/rust/src/packed_array_test.rs

```diff
@@ -176,3 +176,121 @@ fn packed_array_reverse() {
+    let array = PackedByteArray::from(vec.as_slice());
+    assert_eq!(array.as_slice(), vec.as_slice());
+
+    assert_eq!(
+        PackedByteArray::new().as_slice(),
+        &[],
+        "Array is empty, but still a valid slice"
+    );
+}
+
+#[itest]
+fn packed_array_is_mut_unique() {
+    let array1 = PackedByteArray::from(&[1, 2]);
+    let mut array2 = array1.clone();
+    let array3 = array1.to_variant();
+    let array4 = godot::engine::Image::create_from_data(
+        2,
+        1,
+        false,
+        godot::engine::image::Format::FORMAT_R8,
+        array1.clone(),
+    )
+    .unwrap();
+
+    assert_eq!(
+        array1.as_slice().as_ptr(),
+        array2.as_slice().as_ptr(),
+        "Arrays should share the same buffer"
+    );
+    assert_eq!(
+        array1.as_slice().as_ptr(),
+        array3.to::<PackedByteArray>().as_slice().as_ptr(),
+        "Arrays should share the same buffer. Even when stored in a variant"
+    );
+    assert_eq!(
+        array1.as_slice().as_ptr(),
+        array4.get_data().as_slice().as_ptr(),
+        "Arrays should share the same buffer. Even when stored in an Image"
+    );
+
+    // array2 should become an unique copy, as mutable access triggers copy-on-write.
+    array2.as_mut_slice();
+    assert_ne!(
+        array2.as_slice().as_ptr(),
+        array1.as_slice().as_ptr(),
+        "Arrays should not share the same buffer after a mutable access"
+    );
+    assert_ne!(
+        array2.as_slice().as_ptr(),
+        array3.to::<PackedByteArray>().as_slice().as_ptr(),
+        "Arrays should not share the same buffer after a mutable access. Even when stored in a Variant"
+    );
+    assert_ne!(
+        array2.as_slice().as_ptr(),
+        array4.get_data().as_slice().as_ptr(),
+        "Arrays should not share the same buffer after a mutable access. Even when stored in an Image"
+    );
+
+    assert_eq!(
+        array1.as_slice(),
+        array2.as_slice(),
+        "Array2 should be a copy of array1"
+    );
+
+    // These were not mutably accessed, so they should still share the same buffer.
+    assert_eq!(
+        array1.as_slice().as_ptr(),
+        array3.to::<PackedByteArray>().as_slice().as_ptr(),
+    );
+    assert_eq!(
+        array3.to::<PackedByteArray>().as_slice().as_ptr(),
+        array4.get_data().as_slice().as_ptr(),
+    );
+}
+
+#[derive(GodotClass)]
+#[class(base=RefCounted)]
+struct PackedArrayTest {
+    array: PackedByteArray,
+    _base: Base<RefCounted>,
+}
+
+#[godot_api]
+impl PackedArrayTest {
+    #[func]
+    fn set_array(&mut self, array: PackedByteArray) {
+        self.array = array;
+    }
+
+    #[func]
+    fn get_array(&self) -> PackedByteArray {
+        self.array.clone()
+    }
+
+    #[func]
+    fn are_separate_buffer(&self, other: PackedByteArray) -> bool {
+        self.array.as_slice().as_ptr() != other.as_slice().as_ptr()
+    }
+
+    #[func]
+    fn do_mutable_access(&mut self) {
+        self.array.as_mut_slice();
+    }
+}
```

Comment:
This type is currently acting as an abstraction that makes the tests quite a bit harder to understand, because one array is encapsulated, while the other is not.

I would remove the setter/getter as well as any state (struct fields), and just provide two functions:

1. `is_buffer_shared(&self, lhs: &PackedArray, rhs: &PackedArray)`
    * basically the opposite of `are_separate_buffer`, but doesn't use the double negation
2. `trigger_cow(&self, array: &mut PackedArray)`
    * same as `do_mutable_access`, but expressing the intent better

The receiver `&self` is only necessary because `#[func]` expects it, but you don't access the object's state. You also don't need to have a `#[base]` field or a `impl GodotExt`.

And if you have this abstraction, use it in Rust! 🙂 
(see other comment)